web/modules/custom/solution/src/SolutionAffiliationFieldItemList.php

```diff
@@ -102,55 +127,67 @@ protected function getAffiliation(): array {
+    $connection = $this->getSparqlConnection();
+    $graph_uris = $this->getAvailableGraphs();
+
     /** @var \Drupal\sparql_entity_storage\SparqlEntityStorageFieldHandlerInterface $field_handler */
     $field_handler = \Drupal::service('sparql.field_handler');
     $field_uri = $field_handler->getFieldPredicates('rdf_entity', 'field_ar_affiliates')['collection'];
 
-    // Get existing affiliation.
-    $select[] = 'SELECT ?id';
-    $select[] = "FROM <{$graph_uri}>";
-    $select[] = "WHERE { ?id <{$field_uri}> <{$solution->id()}> . }";
-    $select[] = "ORDER BY (?id)";
-    $existing_ids = [];
-    foreach ($sparql->query(implode("\n", $select)) as $item) {
-      $existing_ids[] = (string) $item->id;
-    }
+    $affected_ids = [];
+    foreach ($graph_uris as $graph_uri) {
+      // Get existing affiliation.
+      $select = [];
+      $select[] = 'SELECT ?id';
+      $select[] = "FROM <{$graph_uri}>";
+      $select[] = 'WHERE {';
+      $select[] = "?id <{$field_uri}> <{$solution->id()}> .";
+      // Ensure the entity exists in the graph.
+      $select[] = "?id a ?type .";
+      $select[] = '}';
+      $select[] = "ORDER BY (?id)";
+
+      $existing_ids = [];
+      foreach ($connection->query(implode("\n", $select)) as $item) {
+        $existing_ids[] = (string) $item->id;
+      }
 
-    $new_ids = array_map(function (EntityReferenceItem $field_item): string {
-      return $field_item->target_id;
-    }, $this->list);
-    sort($new_ids);
+      $new_ids = array_map(function (EntityReferenceItem $field_item): string {
+        return $field_item->target_id;
+      }, $this->list);
+      sort($new_ids);
 
-    // The affiliation has been preserved. Exit here.
-    if ($new_ids === $existing_ids) {
-      return;
-    }
+      // The affiliation has been preserved. Exit here.
+      if ($new_ids === $existing_ids) {
+        return;
+      }
 
-    // Delete existing affiliation.
-    $delete[] = "WITH <{$graph_uri}>";
-    $delete[] = "DELETE { ?id <{$field_uri}> <{$solution->id()}> . }";
-    $delete[] = "WHERE { ?id <{$field_uri}> <{$solution->id()}> . }";
-    $sparql->query(implode("\n", $delete));
-
-    // Insert new affiliation.
-    $insert[] = "WITH <{$graph_uri}>";
-    $insert[] = 'INSERT {';
-    foreach ($new_ids as $id) {
-      $insert[] = "  <{$id}> <{$field_uri}> <{$solution->id()}> .";
+      // Delete existing affiliation.
+      $delete = [];
+      $delete[] = "WITH <{$graph_uri}>";
+      $delete[] = "DELETE { ?id <{$field_uri}> <{$solution->id()}> . }";
+      $delete[] = "WHERE { ?id <{$field_uri}> <{$solution->id()}> . }";
+      $connection->query(implode("\n", $delete));
+
+      // Insert new affiliation.
+      $insert = [];
+      $insert[] = "WITH <{$graph_uri}>";
+      $insert[] = 'INSERT {';
+      foreach ($new_ids as $id) {
+        $insert[] = "  <{$id}> <{$field_uri}> <{$solution->id()}> .";
+      }
+      $insert[] = '}';
+      $connection->query(implode("\n", $insert));
```

Comment:
Same here. I think there's a SPARQL `INSERT` syntax that allows inserting multiple values in multiple graphs. Should be something like:


```
INSERT DATA {
  GRAPH <http://graph1> {
    <a1> <b1> <c1> .
    <a2> <b2> <c2> .
    ...
  }
  GRAPH <http://graph2> {
    <a1> <b1> <c1> .
    <a2> <b2> <c2> .
    ...
  }
  ...
} 
```

Then we use the `foreach (...) {...}` only to construct the query blob and we run a single query after `foreach (...) {...}`.

Reply:
I tried to merge in the two queries but was a failure because the delete query works with a variable for an entity id but in the above way the insert works with a direct uri. So the delete and the insert are separated..